File: MarkdownToPlayground/MarkdownToPlayground/__OBJC/ZFRProcessInfo.m
//
//  ZFRProcessInfo.m
//  MarkdownToPlayground
//
//  Created by Francesco Frison on 07/06/2014.
//  Copyright (c) 2014 Ziofritz. All rights reserved.
//

#import "ZFRProcessInfo.h"

@implementation ZFRProcessInfo


+ (NSDictionary *)processInfoWithFull:(BOOL)isFull
{
    NSArray *keys = @[@"HOME", @"USER", @"PWD"];

    NSDictionary *all = [[NSProcessInfo processInfo] environment];
    if (!isFull) {
        all = [all dictionaryWithValuesForKeys:keys];
    }
    NSMutableDictionary *environment = [[all dictionaryWithValuesForKeys:keys] mutableCopy];
    
    __block NSString *pendingKey = nil;
    __block NSInteger unkeyedIndex = 0;
    NSArray *arguments = [[NSProcessInfo processInfo] arguments];
    [arguments enumerateObjectsUsingBlock:^(NSString *obj, NSUInteger idx, BOOL *stop) {
        if ([obj hasPrefix:@"-"]) {
            pendingKey = [obj stringByReplacingOccurrencesOfString:@"-" withString:@""];
            return;
        }
        
        if (idx == 0) {
            pendingKey = @"build_path";
        }
        
        NSString *key;
        if (pendingKey) {
            key = pendingKey;
            pendingKey = nil;
        }
        else {
            key = [NSString stringWithFormat:@"%ld", unkeyedIndex];
            unkeyedIndex++;
        }
        environment[key] = obj;
    }];
    
    return [environment copy];
}

+ (NSDictionary *)fullProcessInfo
{
    return [self processInfoWithFull:YES];
}

+ (NSDictionary *)processInfo
{
    return [self processInfoWithFull:NO];
}

@end

File: MarkdownToPlayground/MarkdownToPlayground/ZFRProcessInfo.swift
//
//  ZFRProcessInfo.swift
//  MarkdownToPlayground
//
//  Created by Francesco Frison on 15/06/2014.
//

import Foundation

class ZFRProcessInfo {
    
//    var home : String
//    var user: String
//    var pwd: String
    
    class func allInfo() -> [String: String] {
        return info(full: true)
    }
    
    class func info() -> [String: String] {
        return info(full: false)
    }
    
    class func info(#full : Bool) -> [String: String] {
        let keys = ["HOME", "USER", "PWD"]
        var rawEnviroment: [NSObject : AnyObject] = NSProcessInfo.processInfo().environment
        var environment = [String: String]()
        for (key, value) in rawEnviroment {
            let stringKey = key as String
        
            if (!full && !contains(keys, stringKey)) {
                continue
            }
            
            environment[stringKey] = (value as String)
        }
        
        var pendingKey : String?;
        var unkeyedIndex  = 0
        let args = NSProcessInfo.processInfo().arguments
        for  (index, value : AnyObject) in enumerate(args){
            let arg = value as NSString
            if arg.hasPrefix("-") {
                pendingKey = arg.substringFromIndex(1)
                continue
            }
            
            if index == 0 {
                pendingKey = "build_path"
            }
            
            var key : NSString
            if var mutKey : NSString = pendingKey {
                while mutKey.hasPrefix("-") {
                    mutKey = mutKey.substringFromIndex(1)
                }
                key = mutKey
                pendingKey = nil
            }
            else {
                key = "\(unkeyedIndex)"
                unkeyedIndex++
            }
            environment[key as String] = (arg as String)
        }
        
        return environment

    }
}